### PR TITLE
Implement serde for CSV and Parquet FileSinkExec

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/8645

## Rationale for this change

Needed by Ballista, so that we can support `DataFrame:write_xxx` again.

## What changes are included in this PR?

Implement serde for CSV and Parquet FileSinkExec, based on existing support for JSON FileSinkExec.

## Are these changes tested?

Yes, new tests added.

### CODE REVIEW

**Comment by andygrove** [comment]
@devinjdangelo fyi, this is now ready for review

File: datafusion/proto/proto/datafusion.proto [hunk 1]
```diff
@@ -1220,20 +1222,22 @@ message ParquetWriterOptions {
 }
 
 message CsvWriterOptions {
```
Comment (andygrove) [inline datafusion/proto/proto/datafusion.proto:1224]: We have not released a version of DataFusion that contains `CsvWriterOptions` yet, so it is safe to change the field numbers here.

**Review by avantgardnerio — APPROVED** [review]

**Comment by devinjdangelo** [comment]
This LGTM @andygrove! I am planning to attempt a more significant refactor in #8667 as discussed with @alamb and might be able to simplify adding support for some of the more advanced file writing options, including externally defined file types is the goal.

**Comment by andygrove** [comment]
Thanks for the reviews @avantgardnerio and @devinjdangelo

**Pull request merged**